File: opendev/repl/repl.py
"""Interactive REPL for OpenDev."""

import logging
from typing import Optional

logger = logging.getLogger(__name__)

from prompt_toolkit import PromptSession
from prompt_toolkit.history import FileHistory
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.styles import Style
from rich.console import Console
from rich.panel import Panel

from opendev.core.runtime.approval import ApprovalManager
from opendev.core.runtime import (
    ConfigManager,
    ModeManager,
    OperationMode,
)
from opendev.core.context_engineering.history import SessionManager, UndoManager
from opendev.core.runtime.monitoring import ErrorHandler
from opendev.core.runtime.services import RuntimeService
from opendev.core.context_engineering.tools.implementations import (
    FileOperations,
    WriteTool,
    EditTool,
    BashTool,
)
from opendev.ui_textual.components.console_animations import Spinner
from opendev.ui_textual.components import StatusLine, NotificationCenter
from opendev.ui_textual.autocomplete import SwecliCompleter
from opendev.ui_textual.formatters_internal.output_formatter import OutputFormatter
from opendev.ui_textual.style_tokens import (
    CYAN,
    ERROR,
    GREY,
    PT_BG_BLACK,
    PT_BG_SELECTED,
    PT_GREEN,
    PT_GREY,
    PT_META_GREY,
    PT_ORANGE,
    WARNING,
)

# Command handlers
from opendev.repl.commands import (
    SessionCommands,
    ModeCommands,
    MCPCommands,
    HelpCommand,
    ConfigCommands,
    ToolCommands,
    AgentsCommands,
    SkillsCommands,
    PluginsCommands,
    SessionModelCommands,
)

# UI components
from opendev.repl.ui import (
    MessagePrinter,
    InputFrame,
    PromptBuilder,
    Toolbar,
    ContextDisplay,
)

# Query processing
from opendev.repl.query_processor import QueryProcessor


class REPL:
    """Interactive REPL for AI-powered coding assistance."""

    def __init__(
        self,
        config_manager: ConfigManager,
        session_manager: SessionManager,
        *,
        is_tui: bool = False,
    ):
        """Initialize REPL.

        Args:
            config_manager: Configuration manager
            session_manager: Session manager
            is_tui: Whether running inside the TUI (disables interactive prompts)
        """
        self.config_manager = config_manager
        self.session_manager = session_manager
        self.is_tui = is_tui
        self.config = config_manager.get_config()
        self.console = Console()

        # Initialize tools and managers
        self._init_tools()
        self._init_managers()
        self._init_runtime_service()

        # Initialize UI
        self._init_ui_components()
        self._init_prompt_session()

        # Initialize command handlers
        self._init_command_handlers()

        # Initialize query processor
        self._init_query_processor()

        # Initialize hooks system
        self._init_hooks()

        self.running = True

    def _init_tools(self):
        """Initialize file operation and command tools."""
        from opendev.core.context_engineering.tools.implementations import (
            WebFetchTool,
            OpenBrowserTool,
            VLMTool,
            WebScreenshotTool,
        )
        from opendev.core.context_engineering.tools.implementations.web_search_tool import (
            WebSearchTool,
        )
        from opendev.core.context_engineering.tools.implementations.notebook_edit_tool import (
            NotebookEditTool,
        )
        from opendev.core.context_engineering.tools.implementations.ask_user_tool import AskUserTool
        from opendev.core.context_engineering.mcp.manager import MCPManager
        from opendev.core.context_engineering.tools.background_task_manager import (
            BackgroundTaskManager,
        )

        self.task_manager = BackgroundTaskManager(self.config_manager.working_dir)
        self.file_ops = FileOperations(self.config, self.config_manager.working_dir)
        self.write_tool = WriteTool(self.config, self.config_manager.working_dir)
        self.edit_tool = EditTool(self.config, self.config_manager.working_dir)
        self.bash_tool = BashTool(
            self.config, self.config_manager.working_dir, task_manager=self.task_manager
        )
        self.web_fetch_tool = WebFetchTool(self.config, self.config_manager.working_dir)
        self.web_search_tool = WebSearchTool(self.config, self.config_manager.working_dir)
        self.notebook_edit_tool = NotebookEditTool(self.config_manager.working_dir)
        self.ask_user_tool = AskUserTool()  # Uses console fallback
        self.open_browser_tool = OpenBrowserTool(self.config, self.config_manager.working_dir)
        self.vlm_tool = VLMTool(self.config, self.config_manager.working_dir)
        self.web_screenshot_tool = WebScreenshotTool(self.config, self.config_manager.working_dir)
        self.mcp_manager = MCPManager(working_dir=self.config_manager.working_dir)

    def _init_managers(self):
        """Initialize operation managers."""
        from opendev.core.runtime.session_model import SessionModelManager

        self.mode_manager = ModeManager()
        self.approval_manager = ApprovalManager(self.console)
        self.error_handler = ErrorHandler(self.console)
        self.undo_manager = UndoManager(self.config.max_undo_history)
        self.session_model_manager = SessionModelManager(self.config)

    def _init_runtime_service(self):
        """Initialize runtime service with tool registry and agents."""
        self.runtime_service = RuntimeService(self.config_manager, self.mode_manager)
        self.runtime_suite = self.runtime_service.build_suite(
            file_ops=self.file_ops,
            write_tool=self.write_tool,
            edit_tool=self.edit_tool,
            bash_tool=self.bash_tool,
            web_fetch_tool=self.web_fetch_tool,
            web_search_tool=self.web_search_tool,
            notebook_edit_tool=self.notebook_edit_tool,
            ask_user_tool=self.ask_user_tool,
            open_browser_tool=self.open_browser_tool,
            vlm_tool=self.vlm_tool,
            web_screenshot_tool=self.web_screenshot_tool,
            mcp_manager=self.mcp_manager,
        )

        self.tool_registry = self.runtime_suite.tool_registry
        self.normal_agent = self.runtime_suite.agents.normal
        self.agent = self.normal_agent

        # Flag for plan mode request via Shift+Tab
        self._pending_plan_request = False

    def _init_ui_components(self):
        """Initialize UI components and state."""
        # UI Components
        self.spinner = Spinner(self.console)
        self.status_line = StatusLine(self.console)
        self.output_formatter = OutputFormatter(self.console, use_claude_style=True)
        self._notification_center = NotificationCenter(self.console)

        # UI state trackers
        self._last_latency_ms: Optional[int] = None
        self._context_sidebar_visible = False
        self._last_prompt: str = ""
        self._last_operation_summary: str = "—"
        self._last_error: Optional[str] = None
        self._key_bindings = self._build_key_bindings()

        # Message printer and input frame
        self.message_printer = MessagePrinter(self.console)
        self.input_frame = InputFrame(self.console)
        self.prompt_builder = PromptBuilder()
        self.toolbar = Toolbar(self.mode_manager, self.session_manager, self.config)
        self.context_display = ContextDisplay(
            self.console,
            self.mode_manager,
            self.session_manager,
            self.config_manager,
            self._notification_center,
        )

    def _init_prompt_session(self):
        """Initialize prompt session with history and autocomplete."""
        # Setup prompt session with history
        from opendev.core.paths import get_paths

        paths = get_paths(self.config_manager.working_dir)
        history_file = paths.global_history_file
        history_file.parent.mkdir(parents=True, exist_ok=True)

        # Create autocomplete for @ mentions and / commands
        self.completer = SwecliCompleter(working_dir=self.config_manager.working_dir)

        # Elegant autocomplete styling
        autocomplete_style = Style.from_dict(
            {
                "completion-menu": f"bg:{PT_BG_BLACK}",
                "completion-menu.completion": "#FFFFFF",
                "completion-menu.completion.current": f"bg:{PT_BG_SELECTED} #FFFFFF",
                "completion-menu.meta": PT_META_GREY,
                "completion-menu.completion.current.meta": "#A0A0A0",
                "mode-normal": f"bold {PT_ORANGE}",
                "mode-plan": f"bold {PT_GREEN}",
                "toolbar-text": PT_GREY,
            }
        )

        self.prompt_session: PromptSession[str] = PromptSession(
            history=FileHistory(str(history_file)),
            completer=self.completer,
            complete_while_typing=True,
            key_bindings=self._key_bindings,
            style=autocomplete_style,
            bottom_toolbar=self.toolbar.build_tokens,
        )

    def _init_command_handlers(self):
        """Initialize slash command handlers."""
        self.session_commands = SessionCommands(
            self.console,
            self.session_manager,
            self.config_manager,
            session_model_manager=self.session_model_manager,
        )

        self.mode_commands = ModeCommands(
            self.console,
            repl=self,
        )

        self.config_commands = ConfigCommands(
            self.console,
            self.config_manager,
            chat_app=None,  # Will be set by ReplChat
        )
        self.config_commands._session_model_manager = self.session_model_manager

        self.mcp_commands = MCPCommands(
            self.console,
            self.mcp_manager,
            refresh_runtime_callback=self._refresh_runtime_tooling,
            agent=self.agent,
        )

        self.tool_commands = ToolCommands(
            console=self.console,
            repl=self,
        )

        self.help_command = HelpCommand(
            self.console,
            self.mode_manager,
        )

        self.agents_commands = AgentsCommands(
            self.console,
            self.config_manager,
            subagent_manager=(
                self.runtime_suite.subagent_manager
                if hasattr(self.runtime_suite, "subagent_manager")
                else None
            ),
        )

        self.skills_commands = SkillsCommands(
            self.console,
            self.config_manager,
        )

        self.plugins_commands = PluginsCommands(
            self.console,
            self.config_manager,
            is_tui=self.is_tui,
        )

        self.session_model_commands = SessionModelCommands(
            self.console,
            self.config_manager,
            self.session_manager,
            self.session_model_manager,
            rebuild_agents_callback=self.rebuild_agents,
            chat_app=None,  # Will be set by TUI
        )

    def _init_query_processor(self):
        """Initialize query processor for AI interactions."""
        self.query_processor = QueryProcessor(
            self.console,
            self.session_manager,
            self.config,
            self.config_manager,
            self.mode_manager,
            self.file_ops,
            self.output_formatter,
            self.status_line,
            self._print_markdown_message,
        )
        self.query_processor.set_notification_center(self._notification_center)

    def _init_hooks(self):
        """Initialize hooks system from settings.json."""
        from opendev.core.hooks.loader import load_hooks_config
        from opendev.core.hooks.manager import HookManager

        try:
            config = load_hooks_config(self.config_manager.working_dir)
            session_id = (
                self.session_manager.current_session.id
                if self.session_manager.current_session
                else ""
            )
            self._hook_manager = HookManager(
                config=config,
                session_id=session_id,
                cwd=str(self.config_manager.working_dir),
            )
            # Wire into tool registry
            self.tool_registry.set_hook_manager(self._hook_manager)
            # Wire into query processor / react executor
            self.query_processor.set_hook_manager(self._hook_manager)
            # Wire into subagent manager
            if hasattr(self.runtime_suite, "subagent_manager") and self.runtime_suite.subagent_manager:
                self.runtime_suite.subagent_manager.set_hook_manager(self._hook_manager)
            # Wire into compactor (if available on react_executor)
            if hasattr(self.query_processor._react_executor, "_compactor"):
                compactor = self.query_processor._react_executor._compactor
                if compactor and hasattr(compactor, "set_hook_manager"):
                    compactor.set_hook_manager(self._hook_manager)
        except Exception as e:
            logger.warning("Failed to initialize hooks: %s", e)
            self._hook_manager = None

    def _refresh_runtime_tooling(self) -> None:
        """Refresh tool registry and agent metadata after MCP changes."""
        if hasattr(self.tool_registry, "set_mcp_manager"):
            self.tool_registry.set_mcp_manager(self.mcp_manager)
        self.runtime_suite.refresh_agents()

    def rebuild_agents(self) -> None:
        """Rebuild agents with updated configuration (e.g., after model/provider change).

        This is needed when the model provider changes, as the HTTP client
        needs to use the new provider's API key.
        """
        # Refresh config
        self.config = self.config_manager.get_config()

        # Rebuild agents with new config
        self.runtime_suite.rebuild_agents(
            self.config,
            self.mode_manager,
            self.config_manager.working_dir,
        )

        # Update agent references
        self.normal_agent = self.runtime_suite.agents.normal
        self.agent = self.normal_agent

    def _build_key_bindings(self) -> KeyBindings:
        """Configure prompt key bindings for high-speed workflows."""
        kb = KeyBindings()

        @kb.add("s-tab")
        def _(event) -> None:
            from opendev.core.runtime.mode_manager import OperationMode

            if self.mode_manager.current_mode == OperationMode.PLAN:
                # Switch back to normal mode
                self.mode_manager.set_mode(OperationMode.NORMAL)
                self._pending_plan_request = False
                self._notify("Switched to Normal mode.", level="info", toast=False)
            else:
                self._pending_plan_request = not self._pending_plan_request
                if self._pending_plan_request:
                    self._notify(
                        "Plan mode requested. Next query will trigger planning.",
                        level="info",
                        toast=False,
                    )
                else:
                    self._notify("Plan mode cancelled.", level="info", toast=False)

            if hasattr(self, "approval_manager"):
                self.approval_manager.reset_auto_approve()

            event.app.invalidate()

        return kb

    def _print_markdown_message(
        self,
        content: str,
        *,
        symbol: str = "⏺",
    ) -> None:
        """Render assistant content as simple plain text with a leading symbol."""
        self.message_printer.print_markdown_message(content, symbol=symbol)

    def _notify(self, message: str, level: str = "info", *, toast: bool = True) -> None:
        """Record a notification and optionally display a toast."""
        title_map = {
            "info": ("Info", "cyan"),
            "warning": ("Warning", "yellow"),
            "error": ("Error", "red"),
        }
        title, style = title_map.get(level, ("Info", "cyan"))
        self._notification_center.add(level, message)
        if toast:
            self.console.print(Panel(message, title=title, border_style=style, expand=False))

    def _show_notifications(self) -> None:
        """Render the notification center."""
        self.console.print()
        self._notification_center.render()

    def _render_context_overview(self) -> None:
        """Render a compact context sidebar above the prompt."""
        self.context_display.render(
            last_prompt=self._last_prompt,
            last_operation_summary=self._last_operation_summary,
            last_error=self._last_error,
        )

    def start(self, *, startup_type: str = "startup") -> None:
        """Start the REPL loop.

        Args:
            startup_type: How the session started ("startup", "resume", "clear").
        """
        self._print_welcome()

        # Fire SessionStart hook
        if self._hook_manager:
            from opendev.core.hooks.models import HookEvent

            self._hook_manager.run_hooks(
                HookEvent.SESSION_START, match_value=startup_type
            )

        # Connect to enabled MCP servers
        self._connect_mcp_servers()

        # Project instructions (SWECLI.md) are now included in the system prompt
        # via EnvironmentContext, so no separate reminder is needed here.

        while self.running:
            try:
                if self._context_sidebar_visible:
                    self._render_context_overview()

                self.input_frame.print_top()
                user_input = self.prompt_session.prompt(self.prompt_builder.build_tokens())
                self.input_frame.print_bottom()

                if not user_input.strip():
                    continue

                # Check for slash commands
                if user_input.startswith("/"):
                    self._handle_command(user_input)
                    continue

                self._last_prompt = user_input.strip()

                # Process as regular query
                self._process_query(user_input)

            except KeyboardInterrupt:
                self.console.print(f"\n[{WARNING}]Exiting...[/{WARNING}]")
                self.running = False
                break
            except EOFError:
                break

        self._cleanup()

    def _print_welcome(self) -> None:
        """Print compact welcome banner using shared welcome module."""
        from opendev.ui_textual.components import WelcomeMessage

        # Generate welcome content using shared module
        welcome_lines = WelcomeMessage.generate_full_welcome(
            current_mode=self.mode_manager.current_mode,
            working_dir=self.config_manager.working_dir,
        )

        # Print each line with Rich formatting
        for line in welcome_lines:
            # Apply color based on content
            if line.startswith("╔") or line.startswith("║") or line.startswith("╚"):
                self.console.print(f"[white]{line}[/white]")
            elif "Essential Commands:" in line:
                self.console.print(f"[bold white]{line}[/bold white]")
            elif "/help" in line or "/mode" in line:
                styled = line.replace("/help", f"[{CYAN}]/help[/{CYAN}]")
                styled = styled.replace("/mode plan", f"[{CYAN}]/mode plan[/{CYAN}]")
                styled = styled.replace("/mode normal", f"[{CYAN}]/mode normal[/{CYAN}]")
                self.console.print(styled)
            elif "Shortcuts:" in line:
                styled = f"[bold white]{line.split(':')[0]}:[/bold white]"
                rest = line.split(":", 1)[1] if ":" in line else ""
                styled += rest.replace("Shift+Tab", f"[{WARNING}]Shift+Tab[/{WARNING}]")
                styled = styled.replace("@file", f"[{WARNING}]@file[/{WARNING}]")
                styled = styled.replace("↑↓", f"[{WARNING}]↑↓[/{WARNING}]")
                self.console.print(styled)
            elif "Session:" in line:
                mode = self.mode_manager.current_mode.value.upper()
                mode_color = "green" if mode == "PLAN" else "yellow"
                styled = f"[bold white]{line.split(':')[0]}:[/bold white]"
                rest = line.split(":", 1)[1] if ":" in line else ""
                if mode in rest:
                    rest = rest.replace(mode, f"[{mode_color}]{mode}[/{mode_color}]")
                styled += rest
                self.console.print(styled)
            else:
                self.console.print(line)

    def _process_query(self, query: str) -> None:
        """Process a user query with AI using ReAct pattern.

        Args:
            query: User query
        """
        # Check for pending plan request
        plan_requested = self._pending_plan_request
        if plan_requested:
            self._pending_plan_request = False

        # Delegate to query processor
        result = self.query_processor.process_query(
            query,
            self.agent,
            self.tool_registry,
            self.approval_manager,
            self.undo_manager,
            plan_requested=plan_requested,
        )

        # Update state from query processor results
        self._last_operation_summary, self._last_error, self._last_latency_ms = result

    def _process_query_with_callback(self, query: str, ui_callback) -> None:
        """Process a user query with AI using ReAct pattern with UI callback for real-time updates.

        Args:
            query: User query
            ui_callback: UI callback for real-time tool display
        """
        # Check for pending plan request
        plan_requested = self._pending_plan_request
        if plan_requested:
            self._pending_plan_request = False

        # Delegate to query processor with callback
        if hasattr(self.query_processor, "process_query_with_callback"):
            result = self.query_processor.process_query_with_callback(
                query,
                self.agent,
                self.tool_registry,
                self.approval_manager,
                self.undo_manager,
                ui_callback,
                plan_requested=plan_requested,
            )
        else:
            # Fallback to normal processing if callback method doesn't exist
            result = self.query_processor.process_query(
                query,
                self.agent,
                self.tool_registry,
                self.approval_manager,
                self.undo_manager,
                plan_requested=plan_requested,
            )

        # Update state from query processor results
        self._last_operation_summary, self._last_error, self._last_latency_ms = result

    def _handle_command(self, command: str) -> None:
        """Handle slash commands.

        Args:
            command: Command string (including /)
        """
        parts = command.split(maxsplit=1)
        cmd = parts[0].lower()
        args = parts[1] if len(parts) > 1 else ""

        # Route to command handlers
        if cmd == "/help":
            self.help_command.handle(args)
        elif cmd == "/exit" or cmd == "/quit":
            self.running = False
        elif cmd == "/clear":
            self.session_commands.clear()
        elif cmd == "/mode":
            # mode_commands.switch_mode calls mode_manager.set_mode,
            # which triggers the callback to handle agent swapping
            self.mode_commands.switch_mode(args)
        elif cmd == "/models":
            self.config_commands.show_model_selector()
        elif cmd == "/mcp":
            self.mcp_commands.handle(args)
        elif cmd == "/init":
            self.tool_commands.init_codebase(command)
        elif cmd == "/agents":
            self.agents_commands.handle(args)
        elif cmd == "/skills":
            self.skills_commands.handle(args)
        elif cmd == "/plugins":
            self.plugins_commands.handle(args)
        elif cmd == "/session-models":
            self.session_model_commands.handle(args)
        elif cmd == "/compact":
            self.session_commands.compact()
        elif cmd == "/sound":
            from opendev.core.utils.sound import play_finish_sound
            play_finish_sound()
            self.console.print("  ⎿  [cyan]Playing test sound...[/cyan]")
        else:
            self.console.print(f"  ⎿  [{ERROR}]Unknown command[/{ERROR}]")
            self.console.print("  ⎿  Type /help for available commands")
            self.console.print("")  # Blank line for spacing

    def _connect_mcp_servers(self) -> None:
        """Connect to enabled MCP servers on startup asynchronously."""
        import threading

        def connect_in_background():
            """Background thread to connect to MCP servers."""
            try:
                # Get connection results using synchronous wrapper
                results = self.mcp_manager.connect_enabled_servers_sync()

                if results:
                    # Silently connect - no messages
                    self._refresh_runtime_tooling()
            except Exception:
                # Silently fail - user can check with /mcp list
                pass

        # Start connection in background thread - silently
        thread = threading.Thread(target=connect_in_background, daemon=True)
        thread.start()

    def _cleanup(self) -> None:
        """Cleanup resources."""
        # Fire SessionEnd hook
        if self._hook_manager:
            from opendev.core.hooks.models import HookEvent

            try:
                self._hook_manager.run_hooks(HookEvent.SESSION_END)
            except Exception:
                pass  # Best-effort on shutdown
            self._hook_manager.shutdown()

        # Disconnect from MCP servers using manager's shared loop
        try:
            self.mcp_manager.disconnect_all_sync()
        except Exception as e:
            self.console.print(
                f"[{WARNING}]Warning: Error disconnecting MCP servers: {e}[/{WARNING}]"
            )

        # Save current session
        if self.session_manager.current_session:
            self.session_manager.save_session()

        # Display session cost summary on exit
        cost_tracker = getattr(self, "_cost_tracker", None)
        if cost_tracker is None:
            # Try to find it via react_executor
            react_exec = getattr(self, "_react_executor", None)
            if react_exec:
                cost_tracker = getattr(react_exec, "_cost_tracker", None)
        if cost_tracker and cost_tracker.total_cost_usd > 0:
            cost_str = cost_tracker.format_cost()
            tokens = cost_tracker.total_input_tokens + cost_tracker.total_output_tokens
            token_str = f"{tokens / 1000:.1f}K" if tokens >= 1000 else str(tokens)
            self.console.print(
                f"\n[{GREY}]Session cost: {cost_str} | {token_str} tokens | "
                f"{cost_tracker.call_count} API calls[/{GREY}]"
            )

        self.console.print(f"[{CYAN}]Goodbye![/{CYAN}]")

File: crates/opendev-repl/src/repl.rs
//! Main REPL loop: read input -> process -> display.
//!
//! Mirrors `opendev/repl/repl.py`.

use std::io::{self, BufRead, Write};

use tracing::{info, warn};

use opendev_history::SessionManager;
use opendev_runtime::AutonomyLevel;
use opendev_tools_core::ToolRegistry;

use crate::commands::{BuiltinCommands, CommandOutcome};
use crate::error::ReplError;
use crate::query_processor::QueryProcessor;

/// Operation mode for the REPL.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OperationMode {
    /// Normal mode — full tool access.
    Normal,
    /// Plan mode — read-only tools only.
    Plan,
}

impl std::fmt::Display for OperationMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            OperationMode::Normal => write!(f, "NORMAL"),
            OperationMode::Plan => write!(f, "PLAN"),
        }
    }
}

/// State shared across the REPL session.
pub struct ReplState {
    /// Current operation mode.
    pub mode: OperationMode,
    /// Current autonomy level.
    pub autonomy_level: AutonomyLevel,
    /// Whether the REPL is running.
    pub running: bool,
    /// Last user prompt (for context display).
    pub last_prompt: String,
    /// Last operation summary.
    pub last_operation_summary: String,
    /// Last error message (if any).
    pub last_error: Option<String>,
    /// Last LLM latency in milliseconds.
    pub last_latency_ms: Option<u64>,
    /// Whether a plan-mode query is pending (via Shift+Tab toggle).
    pub pending_plan_request: bool,
    /// Flag set by /clear command; REPL loop consumes and clears session messages.
    pub messages_cleared: bool,
    /// Flag set by /compact command; REPL loop consumes and triggers compaction.
    pub compact_requested: bool,
    /// Prompt set by /init command; REPL loop consumes and processes it.
    pub init_prompt: Option<String>,
}

impl Default for ReplState {
    fn default() -> Self {
        Self {
            mode: OperationMode::Normal,
            autonomy_level: AutonomyLevel::default(),
            running: true,
            last_prompt: String::new(),
            last_operation_summary: String::from("—"),
            last_error: None,
            last_latency_ms: None,
            pending_plan_request: false,
            messages_cleared: false,
            compact_requested: false,
            init_prompt: None,
        }
    }
}

/// Interactive REPL for AI-powered coding assistance.
///
/// Orchestrates reading user input, dispatching slash commands,
/// and processing AI queries via the ReAct loop.
pub struct Repl {
    /// Shared REPL state.
    pub state: ReplState,
    /// Session manager for conversation persistence.
    session_manager: SessionManager,
    /// Tool registry for executing tools.
    tool_registry: ToolRegistry,
    /// Query processor for AI interactions.
    query_processor: QueryProcessor,
    /// Built-in command handler.
    commands: BuiltinCommands,
}

impl Repl {
    /// Create a new REPL instance.
    pub fn new(session_manager: SessionManager, tool_registry: ToolRegistry) -> Self {
        let query_processor = QueryProcessor::new();
        let commands = BuiltinCommands::new();
        Self {
            state: ReplState::default(),
            session_manager,
            tool_registry,
            query_processor,
            commands,
        }
    }

    /// Set an initial message to be processed when the REPL starts.
    ///
    /// This message will be processed as a user query before entering
    /// the interactive input loop.
    pub fn set_initial_message(&mut self, message: String) {
        self.state.last_prompt = message;
    }

    /// Run the REPL loop.
    ///
    /// Reads lines from stdin, dispatches commands or processes queries,
    /// and loops until the user exits.
    pub async fn run(&mut self) -> Result<(), ReplError> {
        info!("Starting REPL");
        self.print_welcome();

        // Process initial message if one was set via set_initial_message()
        if !self.state.last_prompt.is_empty() {
            let initial = self.state.last_prompt.clone();
            info!(message = %initial, "Processing initial message");
            self.process_query(&initial).await?;
        }

        let stdin = io::stdin();
        let mut reader = stdin.lock();

        while self.state.running {
            self.print_prompt();

            let mut line = String::new();
            match reader.read_line(&mut line) {
                Ok(0) => {
                    // EOF
                    break;
                }
                Ok(_) => {}
                Err(e) => {
                    warn!(error = %e, "Error reading input");
                    return Err(ReplError::Io(e));
                }
            }

            let input = line.trim();
            if input.is_empty() {
                continue;
            }

            if input.starts_with('/') {
                self.handle_command(input);

                // Consume flags set by commands
                if self.state.messages_cleared {
                    self.state.messages_cleared = false;
                    if let Some(session) = self.session_manager.current_session_mut() {
                        session.messages.clear();
                    }
                }
                if self.state.compact_requested {
                    self.state.compact_requested = false;
                    // Compaction will be driven by ContextCompactor when wired up.
                    info!("Compact flag consumed; compaction will run on next query.");
                }

                if let Some(query) = self.state.init_prompt.take() {
                    self.state.last_prompt = query.clone();
                    self.process_query(&query).await?;
                }

                continue;
            }

            self.state.last_prompt = input.to_string();
            self.process_query(input).await?;
        }

        self.cleanup();
        Ok(())
    }

    /// Print the welcome banner.
    fn print_welcome(&self) {
        println!("OpenDev -- AI-powered coding assistant");
        println!("Type /help for commands, /exit to quit.");
        println!(
            "Mode: {} | Autonomy: {}",
            self.state.mode, self.state.autonomy_level
        );
        println!();
    }

    /// Print the input prompt.
    fn print_prompt(&self) {
        let mode_indicator = match self.state.mode {
            OperationMode::Normal => ">",
            OperationMode::Plan => "plan>",
        };
        print!("{} ", mode_indicator);
        let _ = io::stdout().flush();
    }

    /// Handle a slash command.
    fn handle_command(&mut self, input: &str) {
        let parts: Vec<&str> = input.splitn(2, ' ').collect();
        let cmd = parts[0].to_lowercase();
        let args = parts.get(1).copied().unwrap_or("");

        match self.commands.dispatch(&cmd, args, &mut self.state) {
            CommandOutcome::Handled => {}
            CommandOutcome::Exit => {
                self.state.running = false;
            }
            CommandOutcome::Unknown => {
                eprintln!("Unknown command: {}", cmd);
                eprintln!("Type /help for available commands");
            }
        }
    }

    /// Process a user query through the AI pipeline.
    async fn process_query(&mut self, query: &str) -> Result<(), ReplError> {
        let plan_requested = self.state.pending_plan_request;
        if plan_requested {
            self.state.pending_plan_request = false;
        }

        let result = self
            .query_processor
            .process(
                query,
                &mut self.session_manager,
                &self.tool_registry,
                plan_requested,
            )
            .await?;

        self.state.last_operation_summary = result.operation_summary;
        self.state.last_error = result.error;
        self.state.last_latency_ms = result.latency_ms;

        // Print the assistant response
        if !result.content.is_empty() {
            println!("{}", result.content);
        }

        Ok(())
    }

    /// Clean up resources on exit.
    fn cleanup(&mut self) {
        info!("Cleaning up REPL resources");

        // Persist mode settings into session metadata before saving
        self.session_manager
            .set_metadata("mode", &self.state.mode.to_string());
        self.session_manager
            .set_metadata("autonomy_level", &self.state.autonomy_level.to_string());

        if let Err(e) = self.session_manager.save_current() {
            warn!(error = %e, "Failed to save session on exit");
        }
        println!("Goodbye!");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_operation_mode_display() {
        assert_eq!(OperationMode::Normal.to_string(), "NORMAL");
        assert_eq!(OperationMode::Plan.to_string(), "PLAN");
    }

    #[test]
    fn test_repl_state_default() {
        let state = ReplState::default();
        assert_eq!(state.mode, OperationMode::Normal);
        assert_eq!(state.autonomy_level, AutonomyLevel::SemiAuto);
        assert!(state.running);
        assert!(state.last_prompt.is_empty());
        assert_eq!(state.last_operation_summary, "—");
        assert!(state.last_error.is_none());
        assert!(state.last_latency_ms.is_none());
        assert!(!state.pending_plan_request);
    }

    #[test]
    fn test_autonomy_level_in_state() {
        let mut state = ReplState::default();
        state.autonomy_level = AutonomyLevel::Auto;
        assert_eq!(state.autonomy_level, AutonomyLevel::Auto);
    }
}
